### PR TITLE
Miscellaneous updates and bug fixes

### PR DESCRIPTION
**Pull request by wesleykendall**

### CODE REVIEW

File: pgtransaction/tests/test_atomic.py [hunk 1]
```diff
@@ -88,77 +83,58 @@ class MockError(Exception):
...
+        trade = ddf.G(Trade, company='Coca Cola')
         try:
-            with atomic():
-                Trade.objects.create(company='Coca Cola 2', price=1)
-                raise RuntimeError
-        except RuntimeError:
+            with atomic():  # pragma: no branch
+                trade.id = None
+                trade.save()
+        except IntegrityError:
```
Comment (PaulGilmartin) [inline pgtransaction/tests/test_atomic.py:125]: Any particular reason we switched to IntegrityError here? Main idea of this test was to demonstrate that any exception, not just ones coming from db operations, behaved as expected with respect to nested atomics (i.e. savepoints).

File: pgtransaction/atomic.py [hunk 2]
```diff
@@ -19,32 +18,36 @@ def __init__(
...
-                f"Isolation level {self.isolation_level} not recognised"
-            )
-        if self.isolation_level and self.connection.in_atomic_block:
-            raise PGAtomicConfigurationError(
-                "Setting the isolation level inside in a nested atomic "
-                "transaction is not permitted. Nested atomic transactions "
-                "inherit the isolation level from their parent transaction "
-                "automatically."
-            )
+        if self.isolation_level:
```
Comment (PaulGilmartin) [inline pgtransaction/atomic.py]: This should probably be `if self.isolation_level or self.retry`, since the retry policy only works currently for psycopg2 errors.

File: pgtransaction/atomic.py [hunk 3]
```diff
@@ -18,17 +18,18 @@ def __init__(
...
     ):
         super().__init__(using, savepoint, durable)
-        self.connection = transaction.get_connection(self.using)
         self.isolation_level = isolation_level
         self.retry = retry
         self._used_as_context_manager = True
         self._validate()
 
     def _validate(self):
+        connection = transaction.get_connection(self.using)
```
Comment (PaulGilmartin) [inline pgtransaction/atomic.py]: Great catch, of course when this is a decorator all of this happens at module load time.

File: pgtransaction/tests/test_atomic.py [hunk 4]
```diff
@@ -113,103 +114,148 @@ def test_atomic_retries_context_manager_not_allowed():
...
+    calls = []
+    t1 = threading.Thread(target=concurrent_update, args=[barrier, trade, calls])
+    t2 = threading.Thread(target=concurrent_update, args=[barrier, trade, calls])
+
+    t1.start()
+    t2.start()
+
+    t1.join()
+    t2.join()
+
```
Comment (PaulGilmartin) [inline pgtransaction/tests/test_atomic.py:296]: Really cool you got this to work. I tried pretty much exactly this set up minus the "barrier" thing (which I've never even heard of) and could not get things to work. I wonder barrier was the missing piece.

File: pgtransaction/atomic.py [hunk 5]
```diff
@@ -58,38 +59,40 @@ def __call__(self, func):
...
+                        not in (
                             psycopg2.errors.SerializationFailure,
-                            error,
-                            "",
+                            psycopg2.errors.DeadlockDetected,
                         )
-                    else:
+                        or num_retries >= self.retry
+                    ):
                         raise
```
Comment (PaulGilmartin) [inline pgtransaction/atomic.py:60]: If we raise a SerializationFailure or DeadlockDetected error here, aren't we leaving the data from the transaction where those errors were invoked in an unfinished state? What's rolling back the data correctly if we don't manually invoke __exit__ as was done previously?

**Review by PaulGilmartin — APPROVED** [review]

File: pgtransaction/tests/test_atomic.py (hunk 1 above)
Comment (wesleykendall) [reply pgtransaction/tests/test_atomic.py:125]: I see, will change back!

File: pgtransaction/tests/test_atomic.py (hunk 4 above)
Comment (wesleykendall) [reply pgtransaction/tests/test_atomic.py:296]: Yea the barrier plus the small sleep time makes it pretty reliable

File: pgtransaction/atomic.py (hunk 5 above)
Comment (wesleykendall) [reply pgtransaction/atomic.py:60]: `with self._recreate_cm()` is inside of the try/except, so the transaction is rolled back before we continue the retry loop

File: pgtransaction/atomic.py (hunk 3 above)
Comment (wesleykendall) [reply pgtransaction/atomic.py]: Yea it totally crossed my mind too until I saw a failure in a test. Thank goodness for the test suite!

File: pgtransaction/atomic.py (hunk 2 above)
Comment (wesleykendall) [reply pgtransaction/atomic.py]: I'd argue against checking that `retry` is set. Although we currently only configure it to only catch psycopg2 errors, the code will still work if you switch out the DB as long as psycopg2 is installed. It just won't actually retry anything.

For example, say that a user uses sqlite for their test suite and only uses `retry` and never `isolation_level`. Although this might be a rare edge case, I like to avoid doing things that might be overly restrictive.

File: pgtransaction/tests/test_atomic.py (hunk 1 above)
Comment (wesleykendall) [reply pgtransaction/tests/test_atomic.py]: Addressed in 4acadc0e7f3a6f368444a63c2a0187b6370c0750

**Comment by wesleykendall** [comment]
@PaulGilmartin addressed comments and put a RuntimeError back in that test you mentioned.

I left two convos unresolved if you have any additional thoughts in case I might be overlooking something. It would be nice if you can double check my logic for the transaction closing. I believe it's fine, and the multi-threaded test should also properly ensure that the transaction is being reverted when retrying

File: pgtransaction/atomic.py (hunk 2 above)
Comment (PaulGilmartin) [reply pgtransaction/atomic.py]: Fair point about the tests. I suppose if we make it clear we only catch postgres errors for now then it should be safe enough.

File: pgtransaction/atomic.py (hunk 5 above)
Comment (PaulGilmartin) [reply pgtransaction/atomic.py:65]: Ahh yes of course. This way makes much more sense than the original.

**Review by PaulGilmartin — APPROVED** [review]
